yamlpatch/patch.go

```diff
@@ -0,0 +1,262 @@
+// Copyright (c) 2021 Palantir Technologies. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package yamlpatch is a library for applying RFC6902 JSON patches to yaml documents.
+// It leverages go-yaml v3's Node type to preserve comments, ordering, and most formatting.
+package yamlpatch
+
+import (
+	"bytes"
+	"encoding/json"
+	"reflect"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+)
+
+var (
+	// defaultIndentSpaces configures the yaml encoder used to marshal patched objects.
+	defaultIndentSpaces = 2
+)
+
+// ApplyIndent the patch to a yaml document provided in originalBytes and return the updated content.
+// A best effort is made to minimize changes outside the patched paths but some whitespace changes are unavoidable.
+func Apply(originalBytes []byte, patch Patch) ([]byte, error) {
+	return ApplyIndent(originalBytes, patch, defaultIndentSpaces)
+}
+
+// ApplyIndent the patch to a yaml document provided in originalBytes and return the updated content.
+// indentSpaces configures the indentation used to marshal patched objects.
+// A best effort is made to minimize changes outside the patched paths but some whitespace changes are unavoidable.
+func ApplyIndent(originalBytes []byte, patch Patch, indentSpaces int) ([]byte, error) {
+	node := new(yaml.Node)
+	if err := yaml.Unmarshal(originalBytes, node); err != nil {
+		return nil, errors.Wrap(err, "unmarshal original yaml bytes")
+	}
+
+	for _, op := range patch {
+		if op.Path == nil {
+			return nil, errors.Errorf("op %s requires a path field", op.Type)
+		}
+		var err error
+		switch op.Type {
+		case OperationAdd:
+			err = patchAdd(node, op.Path, op.Value)
+		case OperationRemove:
+			err = patchRemove(node, op.Path)
+		case OperationReplace:
+			err = patchReplace(node, op.Path, op.Value)
+		case OperationMove:
+			err = patchMove(node, op.Path, op.From)
+		case OperationCopy:
+			err = patchCopy(node, op.Path, op.From)
+		case OperationTest:
+			err = patchTest(node, op.Path, op.Value)
+		default:
+			err = errors.Errorf("unexpected op")
+		}
+		if err != nil {
+			return nil, errors.Wrapf(err, "op %s %s", op.Type, op.Path.String())
+		}
+	}
+	buf := bytes.Buffer{}
+	enc := yaml.NewEncoder(&buf)
```

Comment:
Documentation for `yaml.NewEncoder` states that it should be closed after use to flush all data, so would add deferred close.